test/src/test/java/hudson/cli/CLIEnvVarTest.java

```diff
@@ -0,0 +1,207 @@
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotEquals;
+import static org.junit.Assert.assertThat;
+
+public class CLIEnvVarTest {
+
+    @Rule
+    public JenkinsRule r = new JenkinsRule();
+
+    @Rule
+    public TemporaryFolder tmp = new TemporaryFolder();
+
+    private File home;
+    private File jar;
+
+    @Before
+    public void grabCliJar() throws IOException {
+        home = tmp.newFolder();
+        jar = tmp.newFile("jenkins-cli.jar");
+        FileUtils.copyURLToFile(r.jenkins.getJnlpJars("jenkins-cli.jar").getURL(), jar);
+    }
+
+    @Test
+    public void testJENKINS_URL() throws Exception {
+        // Using -s option
+        assertEquals(0, launch("java",
+                               "-Duser.home=" + home,
+                               "-jar", jar.getAbsolutePath(),
+                               "-s", r.getURL().toString(),
+                               "who-am-i")
+        );
+
+        // Without -s option and without JENKINS_URL
+        assertNotEquals(0, launch("java",
+                                  "-Duser.home=" + home,
+                                  "-jar", jar.getAbsolutePath(),
+                                  "who-am-i")
+        );
+
+        // Without -s option but with JENKINS_URL
+        Map<String, String> envars = new HashMap<>();
+        envars.put("JENKINS_URL", r.getURL().toString());
+        assertEquals(0, launch(envars,
+                               "java",
+                               "-Duser.home=" + home,
+                               "-jar", jar.getAbsolutePath(),
+                               "who-am-i")
+        );
+
+        envars = new HashMap<>();
+        envars.put("JENKINS_URL", "http://invalid-url");
+        assertNotEquals(0, launch(envars,
+                                  "java",
+                                  "-Duser.home=" + home,
+                                  "-jar", jar.getAbsolutePath(),
+                                  "who-am-i")
+        );
+
+        // Override JENKINS_URL with -s option
+        envars = new HashMap<>();
+        envars.put("JENKINS_URL", "http://invalid-url");
+        assertEquals(0, launch(envars,
+                               "java",
+                               "-Duser.home=" + home,
+                               "-jar", jar.getAbsolutePath(),
+                               "-s", r.getURL().toString(),
+                               "who-am-i")
+        );
+    }
+
+    @Test
+    public void testJENKINS_USER_IDandJENKINS_API_TOKEN() throws Exception {
+        r.jenkins.setSecurityRealm(r.createDummySecurityRealm());
+        r.jenkins.setAuthorizationStrategy(new MockAuthorizationStrategy().grant(Jenkins.ADMINISTER).everywhere().to("admin"));
+
+        // Using -auth option
+        try (ByteArrayOutputStream baos = new ByteArrayOutputStream()) {
+            assertEquals(0, launch(Collections.emptyMap(), baos, null,
```

Comment:
I understand now.

Just in case the speed were a concern:
You can change it with: https://dzone.com/articles/how-to-change-environment-variables-in-java

Or you can change your code to get the value from `System.getenv` and if it doesn't exist, from `System.getProperty`. The last one can be set with `System.setProperty` and can be useful with tests.

Reply:
I tried something similar to your proposal, but it only worked if I read the variable in the test method. Reading it from an internal class (the `CLI` class in our case), returned always null, as if the environment variable was not set. And after that, I discovered the `decorateByEnv` method that eased my life!

Regarding the property system, I would prefer to keep the same behaviour than JENKINS_URL has.